pkg/status/kubefed.go

```diff
@@ -0,0 +1,99 @@
+)
+
+// KubefedAttributes required attributes for obtaining kubefed status
+type KubefedAttributes struct {
+	GetClusterFunc func() (*cluster.FedCluster, bool)
+	Period         time.Duration
+	Timeout        time.Duration
+	Threshold      int64
+}
+
+// GetKubefedConditions uses the provided kubefed attributes to determine kubefed status
+func GetKubefedConditions(attrs KubefedAttributes) (*kubefed_v1beta1.KubeFedClusterStatus, error) {
+	// look up cluster connection status
+	fedCluster, ok := attrs.GetClusterFunc()
+	if !ok {
+		notFoundMsg := ErrMsgClusterConnectionNotFound
+		notFoundReason := toolchainv1alpha1.ToolchainStatusClusterConnectionNotFoundReason
+		badClusterStatus := kubefed_v1beta1.KubeFedClusterStatus{
+			Conditions: []kubefed_v1beta1.ClusterCondition{
+				{
+					Type:          kubefed_common.ClusterReady,
+					Status:        corev1.ConditionFalse,
+					Reason:        &notFoundReason,
+					Message:       &notFoundMsg,
+					LastProbeTime: metav1.Now(),
+				},
+			},
+		}
+		return &badClusterStatus, fmt.Errorf(notFoundMsg)
+	}
+	clusterStatus := *fedCluster.ClusterStatus.DeepCopy()
+
+	// check conditions of cluster connection
+	if !kubefed_util.IsClusterReady(fedCluster.ClusterStatus) {
+		return &clusterStatus, fmt.Errorf("the cluster connection is not ready")
+	}
+
+	var lastProbeTime metav1.Time
+	foundLastProbeTime := false
+	for _, condition := range fedCluster.ClusterStatus.Conditions {
+		if condition.Type == kubefed_common.ClusterReady {
+			lastProbeTime = condition.LastProbeTime
+			foundLastProbeTime = true
+		}
+	}
+	if !foundLastProbeTime {
+		return &clusterStatus, fmt.Errorf("the time of the last probe could not be determined")
+	}
+
+	// check that the last probe time is within limits. It should be less than (period + timeout) * threshold
+
+	totalf := (attrs.Period.Seconds() + attrs.Timeout.Seconds()) * float64(attrs.Threshold)
+	maxDuration, err := time.ParseDuration(fmt.Sprintf("%fs", totalf))
+	if err != nil {
+		return &clusterStatus, errs.Wrap(err, "the maximum duration since the last probe could not be determined - check the configured values for the kubefed health check period, timeout and failure threshold")
+	}
+
+	lastProbedTimePlusMaxDuration := lastProbeTime.Add(maxDuration)
+	currentTime := time.Now()
+	if currentTime.After(lastProbedTimePlusMaxDuration) {
+		errMsg := fmt.Sprintf("%s: %s", ErrMsgClusterConnectionLastProbeTimeExceeded, maxDuration.String())
+		errReason := toolchainv1alpha1.ToolchainStatusClusterConnectionLastProbeTimeExceededReason
+		badProbeCondition := kubefed_v1beta1.KubeFedClusterStatus{
+			Conditions: []kubefed_v1beta1.ClusterCondition{
+				{
+					Type:          kubefed_common.ClusterReady,
+					Status:        corev1.ConditionFalse,
+					Reason:        &errReason,
+					Message:       &errMsg,
+					LastProbeTime: lastProbeTime,
+				},
+			},
+		}
+		return &badProbeCondition, fmt.Errorf(errMsg)
```

Comment:
same question here: do we need to return an error if there's a timeout, or just a specific condition (as you've done)?

Reply:
As I was making the changes for deployment conditions and kubefed conditions I realized we can improve on how the kubefed conditions are being handled in general. I think it's worth discussing and handling as a separate PR since it does appear to cause some API breaking changes. I opened a separate issue for that work: https://issues.redhat.com/browse/CRT-729

Reply:
ok, sure, that was not a blocker for this PR so addressing it later is fine ;) Thanks!